Source/WebCore/Modules/permissions/PermissionStatus.cpp

```diff
@@ -31,48 +31,97 @@
+    return ++currentIdentifier;
+}
+
+static HashMap<uint64_t, std::unique_ptr<MainThreadPermissionObserver>>& allMainThreadPermissionObservers()
+{
+    static NeverDestroyed<HashMap<uint64_t, std::unique_ptr<MainThreadPermissionObserver>>> map;
+    return map;
+}
+
+static void addMainThreadPermissionObserver(uint64_t identifier, std::unique_ptr<MainThreadPermissionObserver>&& observer)
+{
+    allMainThreadPermissionObservers().add(identifier, WTFMove(observer));
+}
+
+static void removeMainThreadPermissionObserver(uint64_t identifier)
+{
+    allMainThreadPermissionObservers().remove(identifier);
+}
+
+Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, Page* page)
+{
+    auto status = adoptRef(*new PermissionStatus(context, state, descriptor, source, page));
+    if (source != PermissionQuerySource::Window)
+        status->createAssociatedMainThreadPermissionObserver(context);
     status->suspendIfNeeded();
     return status;
 }
 
-PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, Page* page)
     : ActiveDOMObject(&context)
     , m_state(state)
     , m_descriptor(descriptor)
+    , m_source(source)
+    , m_page(page)
 {
     auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
     m_origin = ClientOrigin { context.topOrigin().data(), WTFMove(originData) };
     
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
+    m_mainThreadPermissionObserverIdentifier = 0;
+    if (source == PermissionQuerySource::Window) {
         PermissionController::shared().addObserver(*this);
+        return;
+    }
+
+    m_mainThreadPermissionObserverIdentifier = nextMainThreadPermissionObserverIdentifier();
+}
+
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, protectedThis = Ref { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() {
+        addMainThreadPermissionObserver(m_mainThreadPermissionObserverIdentifier, makeUnique<MainThreadPermissionObserver>(this, contextIdentifier, m_descriptor, origin, m_source, m_page.get(), m_state));
+    });    
 }
 
 PermissionStatus::~PermissionStatus()
 {
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
-        PermissionController::shared().removeObserver(*this);
+    callOnMainThread([identifier = m_mainThreadPermissionObserverIdentifier]() {
```

Comment:
We should not dispatch to the main thread if m_mainThreadPermissionObserverIdentifier is 0.